packages/tokens/non-fungible/src/extensions/consecutive/storage.rs

```diff
@@ -0,0 +1,341 @@
+    NonFungibleTokenError, TokenId,
+};
+
+/// Storage keys for the data associated with `FungibleToken`
+#[contracttype]
+pub enum StorageKey {
+    Approval(TokenId),
+    Owner(TokenId),
+    BurnedToken(TokenId),
+}
+
+// ################## QUERY STATE ##################
+
+/// Returns the address of the owner of the given `token_id`.
+///
+/// # Arguments
+///
+/// * `e` - Access to the Soroban environment.
+/// * `token_id` - Token id as a number.
+///
+/// # Errors
+///
+/// * [`NonFungibleTokenError::NonExistentToken`] - Occurs if the provided
+///   `token_id` does not exist.
+pub fn owner_of(e: &Env, token_id: TokenId) -> Address {
+    let max = sequential::next_token_id(e);
+    let is_burned =
+        e.storage().persistent().get(&StorageKey::BurnedToken(token_id)).unwrap_or(false);
+
+    if token_id >= max || is_burned {
+        panic_with_error!(&e, NonFungibleTokenError::NonExistentToken);
+    }
+
+    //e.storage().persistent().extend_ttl(&key, OWNER_TTL_THRESHOLD,
+    // OWNER_EXTEND_AMOUNT);
+    (0..=token_id)
+        .rev()
+        .map(StorageKey::Owner)
+        // after the Protocol 23 upgrade, storage read cost is marginal,
+        // making the consecutive storage reads justifiable
+        .find_map(|key| e.storage().persistent().get::<_, Address>(&key))
+        .unwrap_or_else(|| panic_with_error!(&e, NonFungibleTokenError::NonExistentToken))
+}
+
+// ################## CHANGE STATE ##################
+
+/// Mints a batch of tokens with consecutive ids and attributes them to `to`.
+/// This function does NOT handle authorization.
+///
+/// # Arguments
+///
+/// * `e` - Access to the Soroban environment.
+/// * `to` - The address of the recipient.
+/// * `amount` - The number of tokens to mint.
+///
+/// # Errors
+///
+/// * refer to [`crate::storage::increase_balance`] errors.
+///
+/// # Events
+///
+/// * topics - `["consecutive_mint", to: Address]`
+/// * data - `[from_token_id: TokenId, to_token_id: TokenId]`
+///
+/// # Security Warning
+///
+/// **IMPORTANT**: The function intentionally lacks authorization controls. You
+/// MUST invoke it only from the constructor or implement proper authorization
+/// in the calling function. For example:
+///
+/// ```ignore,rust
+/// fn mint_batch(e: &Env, to: &Address, amount: TokenId) {
+///     // 1. Verify admin has minting privileges (optional)
+///     let admin = e.storage().instance().get(&ADMIN_KEY).unwrap();
+///     admin.require_auth();
+///
+///     // 2. Only then call the actual mint function
+///     crate::consecutive::batch_mint(e, &to, amount);
+/// }
+/// ```
+///
+/// Failing to add proper authorization could allow anyone to mint tokens!
+pub fn batch_mint(e: &Env, to: &Address, amount: TokenId) -> TokenId {
+    let next_id = sequential::increment_token_id(e, amount);
+
+    e.storage().persistent().set(&StorageKey::Owner(next_id), &to);
+
+    increase_balance(e, to, amount);
+
+    let last_id = next_id + amount - 1;
```

Comment:
I'm confused on why do we add `amount` once more? 
`next_id` is already = `current_id` + `amount`
so, in this code, `last_id` = `current_id` + `amount` + `amount` - 1

Reply:
> `next_id` is already = `current_id` + `amount`

No `next_id` is the first "free" ID, it's basically `first_id` from the batch. Good point, will rename it to avoid confusion.